### PR TITLE
GEODE-5209: Adding GCC code coverage

### PR DESCRIPTION
**Pull request by mcmellawatt**
Co-authored-by: Ryan McMahon <rmcmahon@pivotal.io>
Co-authored-by: Mike Martell <mmartell@pivotal.io>
Co-authored-by: Ernest Burghardt <eburghardt@pivotal.io>

### CODE REVIEW

**Comment by mcmellawatt** [comment]
Need to look into excluding unwanted results (boost, etc)

**Comment by mcmellawatt** [comment]
Excluding unwanted coverage is done when you invoke lcov, which is not part of this PR

File: CMakeLists.txt [hunk 1]
```diff
@@ -256,7 +257,7 @@ elseif(CMAKE_CXX_COMPILER_ID MATCHES "MSVC")
 
   # Enables multiprocess compiles
   set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} /MP")
-  
+
```
Comment (pivotal-jbarrett) [inline CMakeLists.txt]: Can we avoid whitespace changes in the future.

File: cppcache/integration-test/CMakeLists.txt [hunk 2]
```diff
@@ -81,6 +81,12 @@ foreach(FILE ${SOURCES})
   add_executable(${TEST} ${FILE})
   set_target_properties(${TEST} PROPERTIES FOLDER test/cppTests)
 
+  if(${CMAKE_CXX_COMPILER_ID} MATCHES "GNU" AND ${USE_CPP_COVERAGE})
+      set_target_properties(${TEST}
+          PROPERTIES APPEND_STRING COMPILE_FLAGS "--coverage"
+                     APPEND_STRING LINK_FLAGS "--coverage")
```
Comment (pivotal-jbarrett) [inline cppcache/integration-test/CMakeLists.txt]: Are we sure that integration tests have to be compiled with coverage? Would it be sufficient to just link?

File: cppcache/shared/CMakeLists.txt [hunk 3]
```diff
@@ -47,6 +47,12 @@ set_target_properties(apache-geode PROPERTIES
   CXX_VISIBILITY_PRESET hidden
   VISIBILITY_INLINES_HIDDEN ON)
 
+if(${CMAKE_CXX_COMPILER_ID} MATCHES "GNU" AND ${USE_CPP_COVERAGE})
```
Comment (pivotal-jbarrett) [inline cppcache/shared/CMakeLists.txt]: Rather than duplicating these settings in each target how about using an `INTERFACE` target?

In root CMakeLists.txt:
```
add_library(INTERFACE _Coverage)
if(${CMAKE_CXX_COMPILER_ID} MATCHES "GNU" AND ${USE_CPP_COVERAGE})
  set_target_properties(_Coverage PROPERTIES
    COMPILE_FLAGS "--coverage"
    LINK_FLAGS "--coverage")
 endif()
```

In other CMakeLists:
```
target_link_libraries(<target>
  ...
  PRIVATE
  ...
  _Coverage)
```

The bonus is that if we add support for other compilers and linkers we only have to change the root CMakeLists.txt. The other will inherit the coverage flags from he `_Coverage` interface target.

File: cppcache/static/CMakeLists.txt [hunk 4]
```diff
@@ -16,21 +16,21 @@
 cmake_minimum_required(VERSION 3.4)
 project(apache-geode-static LANGUAGES CXX)
 
-add_library(apache-geode-static STATIC ${SOURCES_ALL})
+add_library(${PROJECT_NAME} STATIC ${SOURCES_ALL})
```
Comment (pivotal-jbarrett) [inline cppcache/static/CMakeLists.txt]: Don't use ${PROJECT_NAME} like this. It breaks integration with CLion and other CMake integrations. Revert these changes.

File: CMakeLists.txt (hunk 1 above)
Comment (mcmellawatt) [reply CMakeLists.txt]: We'll watch out for this next time

File: cppcache/integration-test/CMakeLists.txt (hunk 2 above)
Comment (mcmellawatt) [reply cppcache/integration-test/CMakeLists.txt]: The lcov docs indicate that you need both:
http://ltp.sourceforge.net/coverage/lcov.php

File: cppcache/shared/CMakeLists.txt (hunk 3 above)
Comment (mcmellawatt) [reply cppcache/shared/CMakeLists.txt]: Was able to implement that and I like that it reduces duplicated CMake code.

File: cppcache/static/CMakeLists.txt (hunk 4 above)
Comment (mcmellawatt) [reply cppcache/static/CMakeLists.txt]: Done
